### PR TITLE
hc/92-recent-standards

### PR DESCRIPTION
**Pull request by xHayden**
## 92-recent-standards

Issue Number(s): #92 

Adds recentStandards property to User which is updated when a user views standards. These are then displayed next to standard card components on plan builder pages with granular control for the number of standards to see recency of. The request itself is an API call that also has that control.

### Checklist

**Frontend**
- [x] recent standards component created
- [x] recent standards component takes in a prop that specifies how many standards to track + show
- [x] track default 3 standards at the moment
       - the prop should allow you to track more than three standards but default it to 3
       - consider using a queue both for the frontend state and backend database (add to front + remove from back) (see other notes)
- [x] the appropriate standard card component is rendered in the recent section and is scrollable (overflow: scroll)
- [x] onClick on standard opens up the modal for that specific standard (the `CardModal` component) (see other notes)
- [x] onClick on "see all standards" routes back to `/library` page

**Backend**
- [x] array of type card is added to user schema (call it something like "recentStandardsQueue")
- [x] onClick of each standard in the library page, it will update the queue both in the frontend and backend
- [ ] once max number of standards to be displayed is reached (based on the prop), remove from end of array

### How to Test

- Log in
- Click on a few cards in the library. Have fun with it.
- Go to Project Plan Builder (The only place that loads RecentStandardsView in this PR)
- Go to saved plans.
- See if the ones you picked are loaded in the right order. First to last. 
- Click See All Standards. It should take you back to the library page.

### Notes

SWR updated to 2.0 to work with post requests.

StandardCard on open updates the users recent standards, but doesn’t really take advantage of useSWRMutation. Caching is unnecessary. Regardless, useSWRMutation is still used because that’s what the issue said to do.

getIds expects a POST request, not a GET request. This is because the number of ids can be quite large.

The recentStandardsComponent is created (RecentStandardsView) but requires recentStandards to be passed to it as is done in saved-project-plans.jsx right now. This can be easily changed to have a useUser hook in the component itself, I just wasn’t sure on how useSWR does caching and figured it was better to be safe.

There is no cap on the number of standards that get stored right now. I don’t see it as a problem because the amount of data that could potentially be added is negligible in comparison to the side of the project, but let me know if that’s a must have. It could require more database queries.

updateRecentStandards in the server-side User actions file requires potentially two queries if the standard is not already a recent standard. I tried merging them but couldn’t figure it out after 2 hours of testing. If you can get it to work, amazing, but it’s wonky because you’re not working on a collection, but a property of an object in a collection.

The recentStandardsView does NOT re-render when a standard is opened on the same page. It does update, however. Reloading shows the change. Good, bad? Idk. That’s a design decision to be made, I guess. To update it, look into revalidating useSWR.

### CODE REVIEW

File: src/actions/Card.js [hunk 1]
```diff
@@ -18,6 +18,28 @@ export const getCards = async () => {
     });
 };
 
+export const getCardsByIds = async (ids) => {
+  return fetch(urls.api.card.getIds, {
+    method: "POST",
```
Comment (nmhossain02) [inline src/actions/Card.js:23]: A "get" request should use the `GET` method. I'd recommend [this](https://stackoverflow.com/questions/35038857/setting-query-string-using-fetch-get-request) for building the URL a bit easier. You might have to change the request handler in `pages/api/card/getIds` a little bit.

File: src/pages/api/card/getIds.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+// @route   GET api/card/getIds
+// @desc    Gets cards by their ids
+// @access  Public
+const handler = async (req, res) => {
+  try {
+    // const userId = req.session.user.id;
+    let cards;
+    if (req.body) {
+      let ids = req.body;
```
Comment (nmhossain02) [inline src/pages/api/card/getIds.js:12]: (just a heads up that if you change this to be a `GET` route, you might have to change this code to use `req.query`)

File: src/pages/api/user/standards/update.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+// @desc    Gets a COUNT amount of recent standards on a user with its ID
+// @access  Public
+const handler = async (req, res) => {
+  try {
+    const { id, count } = req.query;
+
+    const useId = id ? id : req.body.id;
+    const useCount = count ? count : req.body.count;
+
+    const standards = await getRecentStandards(useId, useCount);
```
Comment (nmhossain02) [inline src/pages/api/user/standards/update.js:14]: Work in progress here?

File: src/actions/User.js [hunk 4]
```diff
@@ -90,3 +90,24 @@ export const getCurrentUser = (cookies) => {
       return json.payload;
     });
 };
+
+export const updateRecentStandards = (cardId) =>
+  fetch(urls.api.user.updateRecentStandards, {
+    method: "POST",
```
Comment (nmhossain02) [inline src/actions/User.js:96]: I don't think it's a big deal, but it should technically be a `PATCH` request if you wanna be nitpicky.

File: src/lib/utils/urls.js [hunk 5]
```diff
@@ -12,10 +12,13 @@ export default {
       login: "/api/user/login",
       logout: "/api/user/logout",
       getCurrent: "/api/user",
+      getStandards: "/api/user/standards/get/",
+      updateStandards: "/api/user/standards/update",
```
Comment (nmhossain02) [inline src/lib/utils/urls.js:16]: It'd be slightly more intuitive to have these URLs be accessed with `api.user.standards.get` and `api.user.standards.update`. Very minor things here, but this discrepancy _could_ be why someone in the future has a bad day haha

**Review by nmhossain02 — COMMENTED** [review]
Okay, I can only comment on little things here and there, which aren't really as important getting the core functionality completed. Let me know how things are going on that, and don't worry too much about the feedback so far.

File: src/pages/api/card/getIds.js (hunk 2 above)
Comment (xHayden) [reply src/pages/api/card/getIds.js:12]: Meant to rename this to POST, the number of ids that should be accepted by the endpoint exceeds the data limit of a GET request. Do you think I should rename the file?

File: src/actions/Card.js (hunk 1 above)
Comment (xHayden) [reply src/actions/Card.js:23]: Meant to rename this getIds to be a POST method, the number of ids that should be accepted by the endpoint exceeds the data limit of a GET request.

**Pull request merged**